crates/core/src/hash.rs

```diff
@@ -51,12 +51,51 @@ pub type HashResult<T> = std::result::Result<T, Error>;
     BorshDeserialize,
     BorshDeserializer,
     BorshSchema,
-    Serialize,
-    Deserialize,
 )]
 /// A hash, typically a sha-2 hash of a tx
 pub struct Hash(pub [u8; HASH_LENGTH]);
 
+impl Serialize for Hash {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        // Convert the byte array to a hex string
+        let hex_string = HEXUPPER.encode(&self.0);
+        // Serialize the hex string
+        serializer.serialize_str(&hex_string)
+    }
+}
+
+impl<'de> Deserialize<'de> for Hash {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        // Deserialize a hex string
+        let hex_string =
+            <String as serde::Deserialize>::deserialize(deserializer)?;
+        // Convert the hex string back to a byte array
+        let bytes = HEXUPPER
+            .decode(hex_string.as_bytes())
+            .map_err(de::Error::custom)?;
+
+        // Ensure the byte array has the correct length
+        if bytes.len() != HASH_LENGTH {
+            return Err(de::Error::custom(format!(
+                "Invalid length: expected {} bytes, got {}",
+                HASH_LENGTH,
+                bytes.len()
+            )));
+        }
+
+        let mut array = [0u8; HASH_LENGTH];
+        array.copy_from_slice(&bytes);
+
+        Ok(Hash(array))
```

Comment:
you can avoid allocating an intermediate buffer
```suggestion
        // Ensure the byte array has the correct length
        if hex_string.len() != 2 * HASH_LENGTH {
            return Err(de::Error::custom(format!(
                "Invalid length: expected {} bytes, got {}",
                2 * HASH_LENGTH,
                hex_string.len()
            )));
        }

        // Convert the hex string back to a byte array
        let mut output = [0u8; HASH_LENGTH];
        HEXUPPER
            .decode_mut(hex_string.as_bytes(), &mut output)
            .expect("Hash decoding shouldn't fail after assertion");
            
        Ok(Hash(output))
```